drivers/hmis/app/graphql/mutations/bulk_assign_service.rb

```diff
@@ -0,0 +1,104 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module Mutations
+  class BulkAssignService < CleanBaseMutation
+    description 'Assign services for a set of Clients. If any client is not enrolled, the client will be enrolled in the project as well.'
+    argument :input, Types::HmisSchema::BulkAssignServiceInput, required: true
+    field :success, Boolean, null: true
+
+    def resolve(input:)
+      project = Hmis::Hud::Project.viewable_by(current_user).find(input.project_id)
+      raise 'unauthorized' unless current_permission?(permission: :can_edit_enrollments, entity: project)
+
+      clients = Hmis::Hud::Client.viewable_by(current_user).where(id: input.client_ids).preload(:enrollments)
+      raise 'clients not found' unless clients.count == input.client_ids.uniq.length
+
+      cst = Hmis::Hud::CustomServiceType.find(input.service_type_id)
+      hud_user_id = Hmis::Hud::User.from_user(current_user).user_id
+
+      # Evaluate perm, but don't raise an exception unless the operation actually needs to perform an enrollment
+      can_enroll_clients = current_permission?(permission: :can_enroll_clients, entity: project)
+
+      # Determine and validate CoC Code, which is needed for creating new Enrollments
+      coc_code = determine_coc_code(coc_code_arg: input.coc_code, project: project)
+
+      project_has_units = project.units.exists?
+      available_units = project.units.unoccupied_on(input.date_provided).order(updated_at: :desc).to_a
+
+      Hmis::Hud::Service.transaction do
+        clients.each do |client|
+          # Look for Enrollment at the project that is open on the service date
+          enrollment = client.enrollments.
+            open_on_date(input.date_provided).
+            with_project(project.id).
+            order(entry_date: :desc, date_created: :asc).last # Tie-break: newest entry date; oldest record
```

Comment:
This logic is duplicated right now with the `active_enrollment` field on Client. Need to either:
1) consolidate logic, 
2) change it so the frontend is passing enrollment id for some clients and client id for other clients. I didn't like this approach, imo cleaning for the backend to handle that, but could be swayed

Reply:
Hm, actually I couldn't do that because now the shared logic differs in implementation. One uses normal AR and one uses data loaders, and the data loader functions are not available in the mutation code. Should they be? I can see this being something that comes up in the future as well -- wanting to share lookup code between query and mutations.

Reply:
I don't think there's any technical reason not to include the data loader helpers in the base mutation. I've never tried calling a loader from within a mutation but I assume it would work. If it does, we could extract the helpers out of BaseMutation into a new module that could be included in BaseMutation and BaseObject.

Or we could try and unify all our helpers into GraphqlApplicationHelper but that might be a larger change

Reply:
Cool. It does work. I moved shared code into `GraphqlApplicationHelper`. I think it is a lot cleaner! There is some code churn, but for the best imo. Commit: https://github.com/greenriver/hmis-warehouse/pull/4087/commits/6004847fe4e406caa7df66faa997b10c0aa64efd

I also removed `default_create_record` helper. We had been waiting to remove it until service creation was removed, looks like there was 1 more usage (`CreateProjectConfig`) but I refactored it.

Reply:
actually - rspec says otherwise, something might be off with the data loading .Looking.

Reply:
PHEW that was a doozy. When I started using the loader code in the mutation, I noticed some weird behavior with how the `scope:` filter worked. I switched to a simpler approach as you had originally suggested:

```
  def load_open_enrollment_for_client(client, project_id:, open_on_date:)
    # Load all visible enrollments for the client
    enrollments = load_ar_association(
      client,
      :enrollments,
      scope: Hmis::Hud::Enrollment.viewable_by(current_user).preload(:client_project, :exit),
    )

    # Filter down by project and date
    enrollments.filter do |en|
      en.open_on_date?(open_on_date) && en.client_project.project_id.to_s == project_id.to_s
    end.min_by { |e| [e.entry_date, e.id] }
  end
```

The preload does indeed work, which is great. Confirmed with the n+1 test case.